katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/QueryBuilder.java

```diff
@@ -0,0 +1,87 @@
+package io.katharsis.jpa.internal.query;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.katharsis.jpa.internal.meta.MetaAttribute;
+import io.katharsis.jpa.internal.meta.MetaDataObject;
+import io.katharsis.jpa.internal.meta.MetaEntity;
+import io.katharsis.jpa.internal.meta.MetaKey;
+import io.katharsis.jpa.internal.query.backend.JpaQueryBackend;
+import io.katharsis.queryspec.FilterSpec;
+
+public class QueryBuilder<T, F, O, P, E> {
+
+	private static final String SORT_COLUMN_ALIAS_PREFIX = "__sort";
+
+	private JpaQueryBackend<F, O, P, E> backend;
+
+	private AbstractJpaQueryImpl<T, ?> query;
+
+	public QueryBuilder(AbstractJpaQueryImpl<T, ?> query, JpaQueryBackend<F, O, P, E> backend) {
+		this.query = query;
+		this.backend = backend;
+	}
+
+	/**
+	 * Returns true if distinct is set manually or the "auto distinct" mode is
+	 * enabled and the user performs a join or fetch on a many assocation. In
+	 * this case, attributes from referenced entities included in the sort
+	 * clause are also added to the select clause.
+	 */
+	protected int applyDistinct() {
+		int numAutoSelections = 0;
+		boolean distinct;
+		if (query.autoDistinct) {
+			// distinct for many join/fetches or manual
+			// in case of ViewQuery we may not need this here, but we need to do
+			// the selection of order-by columns below
+			distinct = query.autoDistinct && !query.autoGroupBy && backend.hasManyRootsFetchesOrJoins();
+			if (distinct) {
+				// we also need to select sorted attributes (for references)
+				int prefixIndex = 0;
+				List<O> newOrderList = new ArrayList<>();
+				for (O order : backend.getOrderList()) {
+					E expression = backend.getExpression(order);
+					if (backend.containsRelation(expression)) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS134)